### ISSUE TITLE
How can i use Label BM-Font ? 

### ISSUE THREAD

**Issue by dalgados** [spritebuilder/SpriteBuilder#98]
hi,

I have two files. first one is font.png and second one is font.fnt. I can drag and drop font.png into spritebuilder successfully but it's not same for font.fnt file. Because SpriteBuilder does not allow to drag and drop font.fnt file. and as a matter of course i can't see my custom font in item properties menu ?

what is the correct way to select custom fonts in sprite builder ?

**Comment by vlidholt** [comment]
We need to figure out how to get this working correctly.

**Comment by vlidholt** [comment]
Strategy for doing this. Create a new format for how to pack bm-fonts. The gm-font needs to be put in a directory with with the '.bmfont' extension. Inside that directory there should be 4 folders, 'resources-phone', 'resources-phonehd', 'resources-tablet' and 'resources-tablethd', each of the directories should include a version of the .fnt file and the .png file. Only the '.bmfont' directory will be visible inside SpriteBuilder.

**Comment by vlidholt** [comment]
This is pretty much implemented now. However, there seems to be a bug in Cocos2d which prevents the font-data/texture to be loaded correctly. What I think is happening is that the texture gets an incorrect content scale (or a content scale that doesn't work with the fnt-file).

It works correctly when the content scale is set to 1 (when displayed as iPhone in SpriteBuilder or as non-retina iPhone in simulator). When displayed on other devices the font doesn't show up (although the correct fnt file and texture seems to be loaded).

Here is a sample bitmap font to experiment with: https://drive.google.com/a/apportable.com/file/d/0BzFSRgoEwrfXclVaUGdNaGNOM2M/edit?usp=sharing

**Comment by vlidholt** [comment]
This issue has now been fixed

**Comment by honkskillet** [comment]
Is this fixed?  I'm using v1.1 and I can't drag and drop a .fnt file into the resources (although I can do so for the .png file).

**Comment by NickyWeber** [comment]
@honkskillet We are working on the next release 1.2, please try a recent commit of the develop branch.

**Comment by LearnCocos2D** [comment]
I just tried this today. I have .fnt and .png files created with GlyphDesigner. After drag & drop onto SB only the PNG file appears in the File View. I renamed the .fnt file to .bmfont as indicated above with the same result.

I tried adding a Label BMFont anyway but the font file selection menu does not offer me a fnt file in the folder where I added the bitmap font.

I'm using the develop branch with commit from 3 days ago (c59e0ac33db4eb032b640bffa4534d04d479d959).

**Comment by LearnCocos2D** [comment]
Btw, which bitmap font formats are officially supported and which tools have support for the format required by SB? For instance Glyph Designer offers me the following export formats:

![screen shot 2014-09-05 at 09 57 25](https://cloud.githubusercontent.com/assets/214599/4161790/5f01e87e-34d2-11e4-9733-93259672d3a1.png)

I tried both GlyphDesigner (.fnt as XML and text) and bmGlyph (cocos2d export). When the font didn't appear with .fnt extension I renamed it to .bmFont and re-added it. It's always the same result: png file shows up, fnt/bmFont is not added to SB.

**Comment by NickyWeber** [comment]
Try `.bmfont`

**Comment by LearnCocos2D** [comment]
I already have, .bmfont or .bmFont made no difference.

Is the .bmfont format supposed to be formatted as XML or text?

**Comment by LearnCocos2D** [comment]
Okay, I tried this again, all of these combinations did not work (files created with GlyphDesigner):
- BMFont XML + .fnt
- BMFont XML + .bmfont
- BMFont XML + .bmFont
- BMFont text + .fnt
- BMFont text + .bmfont
- BMFont text + .bmFont

**Comment by cocojoe** [comment]
I ran into this the other day (1.2) and it honestly took me a few tries to get it working.   I use 'bmGlyph' available on AppStore it includes export options out the box for SpriteBuilder, I would also avoid any kind of mixed case, just keep your font name all lowercase including extension.

Can see result here https://github.com/cocojoe/fontfollow.spritebuilder

**Comment by LearnCocos2D** [comment]
I only have the bmGlyph trial version (same version number as App Store: v2.0.3). This is what the publish dialog looks like:

![screen shot 2014-09-06 at 14 40 24](https://cloud.githubusercontent.com/assets/214599/4175166/04fa49ba-35c3-11e4-9fff-8e821469822e.png)

There is no specific "SpriteBuilder" publishing format. Is that the Cocos2d publish or does the paid version include other publishing options? The cocos2d format generates two files, png + fnt. No subfolders.

Bitmap font import should really work with the default "cocos2d" output from both GD and bmGlyph without user intervention. Currently it doesn't, that's why it seems broken since the two most popular tools' output can't be imported "as is". 

Looking at your project I see that the font is divided into "resources-xxxx" subfolders. That lead me to the following document indicating that subfolders are indeed a requirement: https://github.com/spritebuilder/SpriteBuilder/blob/develop/Documentation/9.%20Using%20Bitmap%20Fonts.md 

At a minimum I would expect dragging a single fnt+png onto SpriteBuilder to create a resources-auto (or resources-phone) folder automatically so that the given bitmap font is used by all resolutions, and the documentation above only needed if you need to import a bitmap font supporting multiple resolutions.

**Comment by cocojoe** [comment]
Click on `Publish Wizard`

**Comment by cocojoe** [comment]
However I do agree the whole situation is not very clear and at the very least should have a tool tip that opens up to relevant documentation on the web. I think this area in general needs a 'refresh' of info.

**Comment by LearnCocos2D** [comment]
Ooooh, the Publish Wizard. Nice, will try this next week.

In case anyone reads this and is wondering: in bmGlyp click Publish, and on that dialog there is a Publish Wizard button (lower left area) that reveals this dialog:

![screen shot 2014-09-06 at 18 04 59](https://cloud.githubusercontent.com/assets/214599/4175640/b14cf4da-35df-11e4-8c8e-266ca3db81be.png)

I asked Mike Daley about the status of SB support in Glyph Designer.

**Comment by NickyWeber** [comment]
If something is going wrong with the wizard have a look at [this](http://forum.spritebuilder.com/t/using-bmglyph-with-spritebuilder-1-1/1447). There is a bug in bmGlyph 2.0.3.

**Comment by LearnCocos2D** [comment]
Good to know! 
I also learned that GlyphDesigner will get SB export support (possibly in v1.9.5) and that you can create SB fonts already by using the GDCL command line tool in a (build) script.

**Comment by LearnCocos2D** [comment]
Btw, what should be the default scaling for SB fonts? Am I correct to assume it should be 100%/2x 50%/25% more often than not, rather than bmGlyph's default of 100%/50%/47%/23%? Since SB defaults to scale up iPads by a factor of exactly 2.0 and not by the "true" aspect ratio.

**Comment by kodeaap** [comment]
Hey @NickyWeber ,

So I've got the latest version of Spritebuilder 1.2.1, and I am trying to use BMFont. 

I've used both products BMGlyph and Glyph Designer to create the 'fontname.bmfont' (all lowercase) folder structure (with 'resources-phone', 'resources-phonehd', 'resources-tablet', 'resources-tablethd' - all containing 'fontname.fnt' and 'fontname.png' files.

I can select the font in Spritebuilder and its displayed perfectly in preview. But once I try to publish it crashes. Is there something I am missing?

Anyone's help who was able to fix this issue would be greatly appreciated. Thank you for your time.

Update:

Well it seems like the rest of the community is also struggling with this issue. I guess I'll rollback to using the cocos2d old school method ('-hd', '-ipad', '-ipadhd').
